### ISSUE TITLE
[Readability Issue] Using __all__ vs. wildcards

### ISSUE THREAD

**Issue by zafartahirov** [tensorflow/tensorflow#5308]
I have noticed that a lot of `__init__.py` files in the repo include 

```python
# pylint: disable=wildcard-import
from <...> import *
# pylint: enable=wildcard-import
```

I believe a more appropriate approach would be to use `__all__`. According to PEP 8 documentation:

> To better support introspection, modules should explicitly declare the names in their public API using the __all__ attribute. Setting __all__ to an empty list indicates that the module has no public API.

**Comment by vrv** [comment]
@drpngx is in the process of sealing our public interface via `__all__` and some other mechanisms.

**Comment by zafartahirov** [comment]
Actually, there is another problem, and I am not sure if I should open a new issue (in case the introduction of `__all__` fixes it) -- it is a little more severe:

A lot of files disable `pylint` messages, but don't enable them in the end -- this potentially might cause a missed warning if several files are imported, because `# pylint disable` is a global disable.

[Here are all the files that modify the `pylint`](https://github.com/tensorflow/tensorflow/search?utf8=%E2%9C%93&q=pylint+disable+wildcard-import&type=Code)
[Here is an example file that doesn't revert the modifications](https://github.com/tensorflow/tensorflow/blob/754048a0453a04a761e112ae5d99c149eb9910dd/tensorflow/contrib/learn/python/learn/preprocessing/__init__.py)

**Comment by vrv** [comment]
Probably a separate issue for that one.

If you want to try to help contribute fixes for those, that would be awesome.

**Comment by drpngx** [comment]
Yes, it's a long story :-)

We are using all, but this has issues, too. So we're starting to roll out `remove_undocumented`.

Here's how it works in a nutshell:

* `__all__` only works one level down
* `remove_undocumented` actually deletes symbols
* Splitting between interface and implementation

So, for instance, in `tensorflow/python/__init_.py`, we do

```
from ...standard_ops import *
from ...platform import app
```

The `__all__` in `standard_ops` works to filter out anything that comes using the `*`. But `app` doesn't, so you still have `tf.app.print_function` and a whole slew of other things.

We're rolling out these changes. Stay tuned.

**Comment by drpngx** [comment]
@zafartahirov nice find, please send a PR fixing the missing ones

**Comment by zafartahirov** [comment]
Yeah, I would love to contribute to the fixes, but I have a paper due the end of the month -- will try to do it over the weekend, no promises though 😄 . If this issue is still open by Saturday, I will send a PR for the current issue as well as #5309

**Comment by drpngx** [comment]
To be clear, I think we have a solution for the `__all__`, but feel free to send a PR for just one file if you have ideas and we can take it from there.

**Comment by zafartahirov** [comment]
I was trying to see how this issue might get resolved without hurting the maintainability and without breaking the package integrity, and nothing comes to mind -- sorry, have nothing to contribute in here.

I think I have found the two solutions that are being "tried" in the [current master](https://github.com/tensorflow/tensorflow/tree/163357b7f3e02ef9b68cd5f107285e27a7f16c8c), I am not sure which one you guys are inclined to use:
1. [`make_all`](https://github.com/tensorflow/tensorflow/blob/163357b7f3e02ef9b68cd5f107285e27a7f16c8c/tensorflow/python/util/all_util.py#L28)
2. resolution in the [`tensorflow/python/__init__.py`](https://github.com/tensorflow/tensorflow/tree/163357b7f3e02ef9b68cd5f107285e27a7f16c8c)

Personally I would just make sure the files that are not supposed to be imported start with `_` to indicate the protection, and use `glob.glob('./[a-zA-Z]*.py')` to get all the public files. After small preprocessing the paths could be converted to the import strings. Protected `_` and private `__` files would be a special case, and would have to be imported/included manually.

The problem with this approach it is non-standard and requires a lot of rewriting. What is the current standard at TF for in-package imports?

**Comment by drpngx** [comment]
Thanks for looking into this!

We want to retain some of the ease and flexibility that we have, while disallowing accidental imports.

Here's what we do:

1. We document symbols that we should use, with `@@foo`
2. We use `remove_undocumented` to delete all symbols from "selected" modules/packages
3. We have a whitelist

"Selected" modules/packages are those that are imported into the `tf` namespace, for instance: `app`, `image`, `nn`, etc., transitively on down. For instance, `tf.app.flags`.

These packages have a choice of (first one chosen to be the predominant solution):

1. Have no implementation, just `from .. import *`, or `from .. import Foo`. They typically have a counterpart, called `_impl`, for instance, `gradients_impl.py`
2. Have implementation, but use underscore for all packages used solely for implementation as you suggest

Here is an example module using the first variant, dubbed dual API/implementation:

`gradients.py`: (API module)

```python
"""Gradients doc
@@some_func
"""
from .. import remove_undocumented
from ..gradients_impl import *
remove_undocumented(__name__, ["some_other_func"])
```

`gradients_impl.py`: (implementation module)

```python

import some_module
from ... import some_useful_module

def some_func(): pass
def some_other_func(): pass
# No remove_undocumented.
```

`foo.py`: (internal implementation module which requires some functions defined in `gradients_impl`. We make sure that `foo.py` is never imported in the `tf` namespace, directly or transitively:

```python

from tensorflow...import gradients_impl

def foo(x):
   return gradients_impl.my_internally_useful_function(x + 1)
```

There are no `import *` in `gradients_impl`.

Here's an example module using the second variant (dubbed the fused api/implementation).

`logging.py`:

```python
"""Logging things
@@print_stderr
"""
import sys as _sys

def print_stderr(msg):
   _sys.stderr.write(msg)

remove_undocumented(__name__, [])
```

To recap, (after the change is fully complete)

1. As a tensorflow user, (minus a handful of exceptions such as `tfdbg`), you are only allowed to use `import tensorflow as tf`
2. As a tensorflow developer, you are encouraged to use `from tensorflow... import ...` in your modules
3. As a tensorflow developer, you are not allowed to use `import *` in your implementation modules, only in API modules
4. So, more or less our internal code remains the same, except that we have to be careful in a handful of modules (perhaps 20 or so)
5. As a tensorflow user, you no longer have access to internal modules that have not been explicitly exported via a whitelist or a documentation reference.

HTH

**Comment by zafartahirov** [comment]
Thank you very much for a detailed explanation. Is this information anywhere in the documentation?

**Comment by drpngx** [comment]
No, we don't have design docs out there. It might be good to have somewhere close to the style guide, though. Ultimately, if you mimic our code, then it should work; if you don't, then we should be able to catch that automatically, at some point.

**Comment by zafartahirov** [comment]
I think we can close this issue due to a) No new comments in the past 2+ months b) This is not really an issue, and the documentation is work in progress.
